### ISSUE TITLE
Setting noErrorTruncation to false truncates inferred type of variables/functions; which are not errors

### ISSUE THREAD

**Issue by AnyhowStep** [microsoft/TypeScript#26238]
**TypeScript Version:**  3.0.1

**Search Terms:** noErrorTruncation

In TS 2.9, if `noErrorTruncation` is unset, or set to `false`, hovering the cursor over an identifier for type inference in VS code would bring up a tooltip with the inferred type.

In TS 3.0, if `noErrorTruncation` is unset, or set to `false`, the tooltip with the inferred type will be truncated with ellipsis.

**Code**

TS 2.9, with `noErrorTruncation : false`

```ts
    /*snip*/
    __canAccept: {
        page?: string | number | null | undefined;
        itemsPerPage?: string | number | null | undefined;
    };
}
```

TS 3.0 with `noErrorTruncation : false`

```ts
    /*snip*/
    __canAccept: {
        ...;
    };
}
```

**Expected behavior:**

`noErrorTruncation` is about truncating errors.

The tooltips displaying inferred types aren't errors, and should not be affected by this flag.

**Actual behavior:**

`noErrorTruncation` affects tooltips displaying inferred types.

-----

If anyone else is having this problem, just go to your `tsconfig.json`, add the following line,

```json
"noErrorTruncation": true,
```

and restart VS code (or your editor of choice)

**Comment by AnyhowStep** [comment]
Weird. I thought I finally "fixed" it but, even with `noErrorTruncation` turned on, some long and complicated types still show up with ellipsis.

Is there a `noInferredTypeTruncation` thing I'm missing?

**Comment by TomaszWaszczyk** [comment]
> 
Weird. I thought I finally "fixed" it but, even with `noErrorTruncation` turned on, some long and complicated types still show up with ellipsis.
> 

The same for me. TypeScript 3.2.2

**Comment by AnyhowStep** [comment]
I'm pretty sure it has something to do with `createElidedInformationPlaceholder`

**Comment by xbtsw** [comment]
The ellipses makes it very hard to debug Pick, Omit, etc, when things went wrong.

**Comment by aleclarson** [comment]
@xbtsw Using `"noErrorTruncation": true` works perfectly on my end. (typescript v3.4.3)

**Comment by xbtsw** [comment]
@aleclarson maybe I should say “when things not went as intended”. As you probably noticed the problems not always manifest itself as an error. Sometimes debugging is done by hovering on an inferred type.

**Comment by AnyhowStep** [comment]
This is kind of blocking my attempts at continuous integration =/

I do a character-by-character comparison of `messageText` for output errors, for my compile-time tests. I remove the working directory from `import("...")` types but the type names keep getting truncated at arbitrary places.

So, different working directory name lengths = different ellipsis positions = failing CI

**Comment by Ciantic** [comment]
It's not just CI, it's usability problem. I actually like browsing (and copying) the type from the hover tip in the VSCode, and when it gets truncated it's pretty much useless.

**Comment by asanacam** [comment]
`noErrorTruncation` also does not solve this problem for me.

**Comment by bradennapier** [comment]
Yes agree here, it would be huge to at least have some tool - perhaps even something dedicated that would allow us to see the fully expanded and normalized type that typescript will use.  Perhaps a custom "peek" style popup for it so that we can choose to it only when we need to in the case of large types that would make processing type extreme.

**Comment by lake2** [comment]
same issue +1
Any way to solve this?

**Comment by xbtsw** [comment]
The ellipsis makes it especially difficult to debug type issues (not type errors) involving `Pick`, `Omit` etc.

For example in a type that involves `Omit` I made a typo in second parameter of `Omit`, then in some other place I wondered why a interface property is still accepted. The ellipsis makes it difficult to trace back and find the original typo.

**Comment by MattiasMartens** [comment]
This is an issue that keeps popping up for me. Any motion on it?

**Comment by Hati-** [comment]
Having this problem as well. Investigating and debugging this issue led me to find this being introduced in merge https://github.com/microsoft/TypeScript/pull/24258.

Typescript starts truncating the output when it reaches the default hard-limit of `160 * 10` characters to prevent the server from hanging too long, even with `"noErrorTruncation": true` set. There's currently no option for VS Code to disable it from the looks of it:
https://github.com/microsoft/TypeScript/blob/57e2fe0462bb897e581aa489f1d6040db559d82b/src/services/utilities.ts#L1993

**Comment by MattiasMartens** [comment]
Thank you for your efforts sleuthing this out @Hati- !

I would hope to see this value bound as a configurable TS preference. However it is awesome to know that there is a way to hack around the limitation.

**Comment by devuxer** [comment]
The hack is good to know about, but it isn't really a solution for me. What I'm running into is that I often want to hover a **type** and want see the entire type (no "..."), but when I'm trying to read an **error**, showing entire type definitions can sometimes make the error utterly unreadable. So, some finer grained control is really needed here. Ideally, there'd be some integration between TypeScript and VSCode to provide a toggle to choose between "verbose" (types expanded) and "compact" (types truncated).

**Comment by mataslib** [comment]
Make this vscode feature please - it's annoying to configure manually.

**Comment by gunters63** [comment]
This can also be improved by setting `"noErrorTruncation": true,` in tsconfig.json without patching Typescripts source.
It will still cut off after some point.

`"noErrorTruncation": false,`: 
![image](https://user-images.githubusercontent.com/90346/208373272-012f3b59-496c-42ca-b8ec-d440e0878f93.png)

`"noErrorTruncation": true,`: 
![image](https://user-images.githubusercontent.com/90346/208373526-0b4f1f4d-55c3-4364-8a1d-3db1e9dc7024.png)

**Comment by vensauro** [comment]
one solution (that are not fast on the project that i works, because needs to load a bunch of things, but works very well) is the extension ts type expand.

https://github.com/d-kimuson/ts-type-expand
![](https://user-images.githubusercontent.com/37296661/119652128-b18edd80-be60-11eb-87b7-aca155ac1210.gif)

**Comment by JohnGemstone** [comment]
> Thanks again for the hack!
> 
> I just wanted to call out that if your VSCode config specifies the `"typescript.tsdk"` option, you'll need to update the `tsserver.ts` in whatever location that option specifies. It would have saved me a bit of debugging time 🙃

Thank you!

**Comment by SpencerKaiser** [comment]
A little more context that I think is worth sharing... _**setting `noErrorTruncation` to `true` can cause some significant performance implications with errors.**_ 

I just spent 20 minutes trying to figure out why an error was popping up in my console but not inline where the error was... went over to GitHub to comment on a few things and let it sit and the error finally popped up and it clicked 🙃

I'm definitely keeping it set to `true` for usability but now have to remember that it can cause debugging issues with really complicated types ☹️

**Comment by DanKaplanSES** [comment]
> I just wanted to call out that if your VSCode config specifies the `"typescript.tsdk"` option, you'll need to update the `tsserver.ts` in whatever location that option specifies. It would have saved me a bit of debugging time 🙃

Thank you for saying this because you saved me a lot of time! Slight correction, though: it's named `tsserver.js`

---

If for nothing else, I thought this was worth reporting for the irony: in my case, the `...` was truncating a three character number.

**Comment by nikelborm** [comment]
There is one awesome extension that is just enormously cool. It allows to explore "compiled" versions of very complex types

Name: TypeScript Explorer
Id: mxsdev.typescript-explorer
Description: Full type information for variables, components, functions, and more in TypeScript projects!
Version: 0.4.0
Publisher: mxs
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=mxsdev.typescript-explorer

**Comment by Zetxus** [comment]
Thanks for the hack @Hati-, 
Another gotcha I found is that if you're using VSCode on the WSL2, your tsserver.js is located here: 

`~/.vscode-server/bin/*SOME_HASH_HERE*/extensions/node_modules/typescript/lib/tsserver.js`

Here's a quick bash script I use to do this change ASAP to avoid any performance issues with high values on the truncation:

```bash
#!/bin/bash

if [ $# -lt 1 ]; then
  echo "Usage: $0 <directory_path> <new_length>?"
  exit 1
fi

directory_path="$1"
new_length="$2"

if [ -z "$2" ]
  then
    echo "No length supplied, will use default (160)"
    new_length=160
fi

file_path=$(find "$directory_path" -name "tsserver.js" 2>/dev/null)
if [ -z "$file_path" ]; then
  echo "tsserver.js file not found within directory $file_path after recursive search"
  exit 1
fi

echo "Found tsserver.js at \"$file_path\""

sed -i "s/var defaultMaximumTruncationLength = [0-9]\+;/var defaultMaximumTruncationLength = $new_length;/" "$file_path"

echo "Length updated successfully: "
cat $file_path | grep "defaultMaximumTruncationLength = "
```

Executing it with just the dir will result in resetting it to the default value of 160:

```sh
changeTypescriptHintLength.sh ~/.vscode-server
```

Executing it with a second argument will change it

```sh
changeTypescriptHintLength.sh ~/.vscode-server 800
```

Note: restarting the TS server is enough for the changes to take effect, no need to restart the entire VSCode.

**Comment by blackhaj** [comment]
This is awesome @Zetxus - thanks so much!

For me, the path name has spaces in it which was upsetting the sed command (I couldn't work out why), so with the help of chatGPT I managed to refactor it to use awk. Sharing here for others:

```
#!/bin/bash

if [ $# -lt 1 ]; then
  echo "Usage: $0 <directory_path> <new_length>?"
  exit 1
fi

directory_path="$1"
new_length="$2"

if [ -z "$2" ]
  then
    echo "No length supplied, will use default (160)"
    new_length=160
fi

file_path=$(find "$directory_path" -name "tsserver.js" 2>/dev/null)
if [ -z "$file_path" ]; then
  echo "tsserver.js file not found within directory $directory_path after recursive search"
  exit 1
fi

echo "Found tsserver.js at \"$file_path\""

# Use awk to perform the replacement
awk -v new_length="$new_length" '{sub(/var defaultMaximumTruncationLength = [0-9]+;/, "var defaultMaximumTruncationLength = " new_length ";")}1' "$file_path" > temp_file && mv temp_file "$file_path"

echo "Length updated successfully: "
cat "$file_path" | grep "defaultMaximumTruncationLength = "

```

**Comment by Djones4822** [comment]
Just to point out, I made this change on my windows install of VS code, I set defaultMaximumTruncationLength=800 and it seem that the change was reverted recently. 

What I don't understand is why typeToString is not respecting the noErrorTruncation compiler option. 

```javascript
  function typeToString(type, enclosingDeclaration, flags = 1048576 /* AllowUniqueESSymbolType */ | 16384 /* UseAliasDefinedOutsideCurrentScope */, writer = createTextWriter("")) {
    const noTruncation = compilerOptions.noErrorTruncation || flags & 1 /* NoTruncation */;
    const typeNode = nodeBuilder.typeToTypeNode(type, enclosingDeclaration, toNodeBuilderFlags(flags) | 70221824 /* IgnoreErrors */ | (noTruncation ? 1 /* NoTruncation */ : 0));
    if (typeNode === void 0)
      return Debug.fail("should always get typenode");
    const printer = type !== unresolvedType ? createPrinterWithRemoveComments() : createPrinterWithDefaults();
    const sourceFile = enclosingDeclaration && getSourceFileOfNode(enclosingDeclaration);
    printer.writeNode(
      4 /* Unspecified */,
      typeNode,
      /*sourceFile*/
      sourceFile,
      writer
    );
    const result = writer.getText();
    const maxLength2 = noTruncation ? noTruncationMaximumTruncationLength * 2 : defaultMaximumTruncationLength * 2;
    if (maxLength2 && result && result.length >= maxLength2) {
      return result.substr(0, maxLength2 - "...".length) + "...";
    }
    return result;
  }
```

something about `const noTruncation = compilerOptions.noErrorTruncation || flags & 1 /* NoTruncation */;` is causing it to get ignored

`noTruncationMaximumTruncationLength` is set to 1e6 so this should not be an issue. I have to assume that when vscode is calling the typeToString function, the flag value is causing the compiler option to be ignored.

**Comment by devcaeg** [comment]
It seems that in recent versions the "defaultMaximumTruncationLength" option is no longer present in the "tsserver.js" file.

**Comment by dbersan** [comment]
> There is one awesome extension that is just enormously cool. It allows to explore "compiled" versions of very complex types
> 
> It allows to explore not only types and interfaces, but literally anything: vars, functions, classes, promises, generics
> 
> Name: TypeScript Explorer Id: mxsdev.typescript-explorer Description: Full type information for variables, components, functions, and more in TypeScript projects! Version: 0.4.0 Publisher: mxs VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=mxsdev.typescript-explorer
> 
> ![Screenshot from 2023-04-23 01-08-07](https://user-images.githubusercontent.com/29357907/233808687-a9de25d4-4af6-471c-9e57-6da412292d36.png)
> 
> ```ts
> function UsePromiseAllMappedTrace<
>   TInputArrayElement,
>   TReturnArrayElement,
> >(
>   arrayToMap: TInputArrayElement[],
>   elementGetter: (
>     currentTraceNode: IPublicActionTraceNode,
>     arrayElement: TInputArrayElement,
>     index: number,
>   ) => Promise<TReturnArrayElement>,
> ): (parentTraceNode: IPublicActionTraceNode) => Promise<TReturnArrayElement[]> {}
> ```

This extension simply doesn't solve the issue. You still need to manually click through this overly complex search tree. I just wanna see the full type, as plain string, it's not that difficult

**Comment by MilesWellsSVT** [comment]
> It seems that in recent versions the "defaultMaximumTruncationLength" option is no longer present in the "tsserver.js" file.
> 
> Where is this option currently available, or where can I increase this value?

Same directory, but now in `typescript.js`

**Comment by aasmal97** [comment]
Also for anyone interested, until we have an in-built fix/alternative, I did find this extension on vscode that might be more stable, and expands the type when you hover over the type.

1. auto-updates to vscode
2. re-installing packages 

will occasionally reset the `defaultMaximunTruncationLength` values in these `.js` files, which will require you to manually update them again. 

Also the most recent version of vscode's appears to sometimes ignore these values, even when set (even after reloading/restarting the server). At least that's what I'm experiencing. 

This is the extension: 
https://marketplace.visualstudio.com/items?itemName=MylesMurphy.prettify-ts
